src/main/java/duke/parser/Parser.java

```diff
@@ -0,0 +1,160 @@
+    private final Ui ui;
+
+    private enum TaskTypes {
+        TODO, DEADLINE, EVENT
+    }
+
+    /**
+     * Creates a Parser that handles user input and turn it into respective
+     * commands and execute the required functions that user demands.
+     *
+     * @param list    The TaskList handler that is handling the list of task.
+     * @param storage The Storage handler that is in-charged of saving and loading files on local directory.
+     * @param ui      The Ui handler that handles printing of output, if required.
+     */
+    public Parser(TaskList list, Storage storage, Ui ui) {
+        this.list = list;
+        this.storage = storage;
+        this.ui = ui;
+    }
+
+    /**
+     * Returns the respective DukeCommand that user want duke to do.
+     * If the command is not supported
+     *
+     * @param userInput User's input on what they want duke to do.
+     * @return The respective command object.
+     * @throws DukeException If some error occurred in the processing of user input,
+     *                       like the user inputted some command that duke does not support.
+     */
+    public DukeCommand processInput(String userInput) throws DukeException {
+        String commandType = userInput.split(" ")[0];
+
+        int spaceIndex = userInput.indexOf(" ");
+        String commandDescription = userInput.substring(spaceIndex + 1);
+        if (!(commandType.equals("list") || commandType.equals("bye"))
+                && (commandDescription.isBlank() || spaceIndex == -1)) {
+            throw new NoCommandDescriptionException();
+        }
+
+        switch (commandType) {
+        case "list":
+            return new List(ui, storage, list);
+
+        case "bye":
+            return new Exit(ui, storage, list);
+
+        case "done":
+            return new Done(ui, storage, list, Integer.parseInt(userInput.split(" ")[1]));
+
+        case "todo":
+            return new Add(ui, storage, list, processTaskDescriptions(TaskTypes.TODO, userInput));
+
+        case "deadline":
+            return new Add(ui, storage, list, processTaskDescriptions(TaskTypes.DEADLINE, userInput));
+
+        case "event":
+            return new Add(ui, storage, list, processTaskDescriptions(TaskTypes.EVENT, userInput));
+
+        case "delete":
+            return new Delete(ui, storage, list, Integer.parseInt(userInput.split(" ")[1]));
+
+        case "find":
+            return new Find(ui, storage, list, userInput.substring(userInput.indexOf(" ") + 1));
+
+        default:
+            throw new UnknownCommandException();
+        }
+    }
+
+    private Task processTaskDescriptions(TaskTypes t, String userInput) throws InvalidDateTimeException, NoTaskDescriptionException, NoDateTimeException {
+        int spaceIndex = userInput.indexOf(" ");
+        String taskDescription = userInput.substring(spaceIndex + 1);
+        if (taskDescription.isBlank() || spaceIndex == -1) {
+            throw new NoTaskDescriptionException();
```

Comment:
Its a bit odd that the processing of task descriptions happen here. While the processing of additional information for other commands happen elsewhere. Perhaps make it consistent and make the processing of the additional information all happen in the commands while the parser only checks for the type of command